### PR TITLE
60 binfhe refactoring

### PR DESCRIPTION
**Pull request by kimandrik**
binfhe folder refactoring:
- moved classes to separate classes
- removed SetQ() calls and the method. Now parameter q is fixed, and all the algorithms rely on input parameters.
- made sk to be initialized mod(qKS) instead of mod(q) (not sure we need it, but made it during removing SetQ())
- There are now two classes RingGSWAccumulatorDM and RingGSWAccumulatorCGGI that extends RingGSWAccumulator and implements accumulation algorithms.

### CODE REVIEW

**Review by yspolyakov — COMMENTED** [review]
Examined the logical changes. All look correct.

Still need to benchmark the bootstrapping after fixing the problem with NATIVE_SIZE=32 and rebased to `dev`.

**Comment by yspolyakov** [comment]
Benchmarking experiments have been run. No issues found. @dsuponitskiy please proceed further with the review.

File: src/binfhe/include/lwe-ciphertext.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+/**
+ * @brief Class that stores a LWE scheme ciphertext; composed of a vector "a"
+ * and integer "b"
+ */
+class LWECiphertextImpl : public Serializable {
+public:
+    LWECiphertextImpl() {}
+
+    explicit LWECiphertextImpl(const NativeVector&& a, const NativeInteger& b) : m_a(std::move(a)), m_b(b) {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-ciphertext.h:59]: do we need explicit for all these constructors?

File: src/binfhe/include/lwe-ciphertext.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+using ConstLWECiphertext = const std::shared_ptr<const LWECiphertextImpl>;
+
+/**
+ * @brief Class that stores a LWE scheme ciphertext; composed of a vector "a"
+ * and integer "b"
+ */
+class LWECiphertextImpl : public Serializable {
+public:
+    LWECiphertextImpl() {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-ciphertext.h:57]: trivial constructor (use "= default")

File: src/binfhe/include/lwe-ciphertext.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    std::string SerializedObjectName() const {
+        return "LWECiphertext";
+    }
+    static uint32_t SerializedVersion() {
+        return 1;
+    }
+
+private:
+    NativeVector m_a;
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-ciphertext.h:166]: m_a and m_b are meaningless names. should be renamed

File: src/binfhe/include/lwe-ciphertext.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    const NativeVector& GetA() const {
+        return m_a;
+    }
+
+    NativeVector& GetA() {
+        return m_a;
+    }
+
+    const NativeInteger& GetA(std::size_t i) const {
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-ciphertext.h:95]: GetA() already exists. The name should be different (maybe GetAByIndex())

File: src/binfhe/include/binfhe-base-params.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _BINFHE_BASE_PARAMS_H_
+#define _BINFHE_BASE_PARAMS_H_
+
+#include <memory>
```
Comment (dsuponitskiy) [inline src/binfhe/include/binfhe-base-params.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/binfhe-base-scheme.h [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef BINFHE_FHEW_H
+#define BINFHE_FHEW_H
+
+#include <map>
```
Comment (dsuponitskiy) [inline src/binfhe/include/binfhe-base-scheme.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/binfhecontext.h [hunk 7]
```diff
@@ -38,72 +38,15 @@
 
 #include <memory>
 #include <string>
-#include <map>
 #include <vector>
+#include <map>
```
Comment (dsuponitskiy) [inline src/binfhe/include/binfhecontext.h:42]: all STL headers should be included after all custom headers.

File: src/binfhe/include/lwe-ciphertext.h [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _LWE_CIPHERTEXT_H_
+#define _LWE_CIPHERTEXT_H_
+
+#include <memory>
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-ciphertext.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/lwe-cryptoparameters.h [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _LWE_CRYPTOPARAMETERS_H_
+#define _LWE_CRYPTOPARAMETERS_H_
+
+#include <string>
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-cryptoparameters.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/lwe-cryptoparameters.h [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+namespace lbcrypto {
+
+/**
+ * @brief Class that stores all parameters for the LWE scheme
+ */
+class LWECryptoParams : public Serializable {
+public:
+    // NativeInteger m_qKS = 1<<20; //PreviousPrime<NativeInteger>(FirstPrime<NativeInteger>(26, 2048), 2048);
+    LWECryptoParams() : m_n(0), m_N(0), m_q(0), m_Q(0), m_qKS(0), m_baseKS(0) {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-cryptoparameters.h:51]: use in-class data member initialization

File: src/binfhe/include/lwe-keyswitchkey.h [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _LWE_KEYSWITCHKEY_H_
+#define _LWE_KEYSWITCHKEY_H_
+
+#include <memory>
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-keyswitchkey.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/lwe-cryptoparameters.h [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        if (Q.GetMSB() > MAX_MODULUS_SIZE) {
+            std::string errMsg = "ERROR: Maximum size of Q supported for FHEW is 60 bits.";
+            OPENFHE_THROW(config_error, errMsg);
+        }
+
+        m_dgg.SetStd(std);
+        m_ks_dgg.SetStd(std);
+    }
+
+    explicit LWECryptoParams(const LWECryptoParams& rhs) {
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-cryptoparameters.h:76]: do we need explicit for constructors here?

File: src/binfhe/include/lwe-cryptoparameters.h [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+            OPENFHE_THROW(deserialize_error, "serialized object version " + std::to_string(version) +
+                                                 " is from a later version of the library");
+        }
+
+        ar(::cereal::make_nvp("n", m_n));
+        ar(::cereal::make_nvp("N", m_N));
+        ar(::cereal::make_nvp("q", m_q));
+        ar(::cereal::make_nvp("Q", m_Q));
+        ar(::cereal::make_nvp("qKS", m_qKS));
+        double sigma;
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-cryptoparameters.h:179]: all variables should be initialized

File: src/binfhe/include/lwe-keyswitchkey.h [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+using ConstLWESwitchingKey = const std::shared_ptr<const LWESwitchingKeyImpl>;
+
+/**
+ * @brief Class that stores the LWE scheme switching key
+ */
+class LWESwitchingKeyImpl : public Serializable {
+public:
+    LWESwitchingKeyImpl() {}
+
+    explicit LWESwitchingKeyImpl(const std::vector<std::vector<std::vector<NativeVector>>>& keyA,
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-keyswitchkey.h:58]: do we need explicit for constructors here?

File: src/binfhe/include/lwe-keyswitchkey.h [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+using LWESwitchingKey = std::shared_ptr<LWESwitchingKeyImpl>;
+
+using ConstLWESwitchingKey = const std::shared_ptr<const LWESwitchingKeyImpl>;
+
+/**
+ * @brief Class that stores the LWE scheme switching key
+ */
+class LWESwitchingKeyImpl : public Serializable {
+public:
+    LWESwitchingKeyImpl() {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-keyswitchkey.h:56]: trivial constructor, use "= default".

File: src/binfhe/include/lwe-keyswitchkey.h [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+    std::string SerializedObjectName() const {
+        return "LWEPrivateKey";
+    }
+    static uint32_t SerializedVersion() {
+        return 1;
+    }
+
+private:
+    std::vector<std::vector<std::vector<NativeVector>>> m_keyA;
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-keyswitchkey.h:133]: add in-class initialization

File: src/binfhe/include/lwe-keyswitchkey.h [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    std::vector<std::vector<std::vector<NativeInteger>>> m_keyB;
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-keyswitchkey.h:134]: add description to m_keyA/B

File: src/binfhe/include/lwe-privatekey.h [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _LWE_PRIVATEKEY_H_
+#define _LWE_PRIVATEKEY_H_
+
+#include <memory>
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-privatekey.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/lwe-privatekey.h [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+using LWEPrivateKey = std::shared_ptr<LWEPrivateKeyImpl>;
+
+using ConstLWEPrivateKey = const std::shared_ptr<const LWEPrivateKeyImpl>;
+
+/**
+ * @brief Class that stores the LWE scheme secret key; contains a vector
+ */
+class LWEPrivateKeyImpl : public Serializable {
+public:
+    LWEPrivateKeyImpl() {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-privatekey.h:56]: trivial constructor

File: src/binfhe/include/lwe-privatekey.h [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+    explicit LWEPrivateKeyImpl(const NativeVector& s) : m_s(s) {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-privatekey.h:58]: explicit for constructors?

File: src/binfhe/include/lwe-privatekey.h [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+    std::string SerializedObjectName() const {
+        return "LWEPrivateKey";
+    }
+    static uint32_t SerializedVersion() {
+        return 1;
+    }
+
+private:
+    NativeVector m_s;
```
Comment (dsuponitskiy) [inline src/binfhe/include/lwe-privatekey.h:125]: add in-class initialization and a comment with description for m_s

File: src/binfhe/include/rgsw-acc-cggi.h [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _RGSW_ACC_CGGI_H_
+#define _RGSW_ACC_CGGI_H_
+
+#include <map>
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acc-cggi.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/rgsw-acc-cggi.h [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace lbcrypto {
+
+/**
+ * @brief Ring GSW accumulator schemes described in
+ * https://eprint.iacr.org/2014/816 and https://eprint.iacr.org/2020/08
+ */
+class RingGSWAccumulatorCGGI : public RingGSWAccumulator {
+public:
+    RingGSWAccumulatorCGGI() {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acc-cggi.h:50]: trivial ctor

File: src/binfhe/include/rgsw-acc-dm.h [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _RGSW_ACC_DM_H_
+#define _RGSW_ACC_DM_H_
+
+#include <map>
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acc-dm.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/rgsw-acc-dm.h [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace lbcrypto {
+
+/**
+ * @brief Ring GSW accumulator schemes described in
+ * https://eprint.iacr.org/2014/816 and https://eprint.iacr.org/2020/08
+ */
+class RingGSWAccumulatorDM : public RingGSWAccumulator {
+public:
+    RingGSWAccumulatorDM() {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acc-dm.h:50]: trivial ctor

File: src/binfhe/include/rgsw-acc.h [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#ifndef _RGSW_FHE_H_
+#define _RGSW_FHE_H_
+
+#include <map>
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acc.h:35]: all STL headers should be included after all custom headers.

File: src/binfhe/include/rgsw-acc.h [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace lbcrypto {
+
+/**
+ * @brief Ring GSW accumulator schemes described in
+ * https://eprint.iacr.org/2014/816 and https://eprint.iacr.org/2020/08
+ */
+class RingGSWAccumulator {
+public:
+    RingGSWAccumulator() {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acc.h:52]: trivial ctor

File: src/binfhe/include/rgsw-acckey.h [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+        m_key.resize(dim1);
+        for (uint32_t i1 = 0; i1 < dim1; i1++) {
+            m_key[i1].resize(dim2);
+            for (uint32_t i2 = 0; i2 < dim2; i2++) {
+                m_key[i1][i2].resize(dim3);
+            }
+        }
+    }
+
+    explicit RingGSWACCKeyImpl(const std::vector<std::vector<std::vector<RingGSWEvalKey>>>& key) : m_key(key) {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acckey.h:80]: explicit?

File: src/binfhe/include/rgsw-acckey.h [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+
+    std::string SerializedObjectName() const {
+        return "RingGSWACCKey";
+    }
+    static uint32_t SerializedVersion() {
+        return 1;
+    }
+
+private:
+    std::vector<std::vector<std::vector<RingGSWEvalKey>>> m_key;
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-acckey.h:146]: initialization missing

File: src/binfhe/include/rgsw-cryptoparameters.h [hunk 30]
```diff
@@ -0,0 +1,318 @@
...
+
+namespace lbcrypto {
+
+/**
+ * @brief Class that stores all parameters for the RingGSW scheme used in
+ * bootstrapping
+ */
+class RingGSWCryptoParams : public Serializable {
+public:
+    RingGSWCryptoParams() : m_N(0), m_Q(0), m_q(0), m_baseG(0), m_digitsG(0), m_baseR(0) {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-cryptoparameters.h:61]: use in-class initialization

File: src/binfhe/include/rgsw-cryptoparameters.h [hunk 31]
```diff
@@ -0,0 +1,318 @@
...
+            OPENFHE_THROW(deserialize_error, "serialized object version " + std::to_string(version) +
+                                                 " is from a later version of the library");
+        }
+        ar(::cereal::make_nvp("bN", m_N));
+        ar(::cereal::make_nvp("bQ", m_Q));
+        ar(::cereal::make_nvp("bq", m_q));
+        ar(::cereal::make_nvp("bR", m_baseR));
+        ar(::cereal::make_nvp("bG", m_baseG));
+        // ar(::cereal::make_nvp("bmethod", m_method));
+        double sigma;
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-cryptoparameters.h:249]: initialize sigma

File: src/binfhe/include/rgsw-cryptoparameters.h [hunk 32]
```diff
@@ -0,0 +1,318 @@
...
+    void Change_BaseG(uint32_t BaseG) {
+        if (m_baseG != BaseG) {
+            m_baseG   = BaseG;
+            m_Gpower  = m_Gpower_map[m_baseG];
+            m_digitsG = (uint32_t)std::ceil(log(m_Q.ConvertToDouble()) / log(static_cast<double>(m_baseG)));
+        }
+    }
+
+private:
+    // ring dimension for RingGSW/RingLWE scheme
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-cryptoparameters.h:272]: in-class initialization

File: src/binfhe/include/rgsw-evalkey.h [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+public:
+    RingGSWEvalKeyImpl() {}
+
+    RingGSWEvalKeyImpl(uint32_t rowSize, uint32_t colSize) {
+        m_elements.resize(rowSize);
+        for (uint32_t i = 0; i < rowSize; i++)
+            m_elements[i].resize(colSize);
+    }
+
+    explicit RingGSWEvalKeyImpl(const std::vector<std::vector<NativePoly>>& elements) : m_elements(elements) {}
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-evalkey.h:75]: explicit?

File: src/binfhe/include/rgsw-evalkey.h [hunk 34]
```diff
@@ -0,0 +1,157 @@
...
+    void SetElements(const std::vector<std::vector<NativePoly>>& elements) {
+        m_elements = elements;
+    }
+
+    /**
+   * Switches between COEFFICIENT and Format::EVALUATION polynomial
+   * representations using NTT
+   */
+    void SetFormat(const Format format) {
+        for (uint32_t i = 0; i < m_elements.size(); i++)
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-evalkey.h:108]: size_t

File: src/binfhe/include/rgsw-evalkey.h [hunk 35]
```diff
@@ -0,0 +1,157 @@
...
+ * @brief Class that stores a RingGSW ciphertext; a two-dimensional vector of
+ * ring elements
+ */
+class RingGSWEvalKeyImpl : public Serializable {
+public:
+    RingGSWEvalKeyImpl() {}
+
+    RingGSWEvalKeyImpl(uint32_t rowSize, uint32_t colSize) {
+        m_elements.resize(rowSize);
+        for (uint32_t i = 0; i < rowSize; i++)
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-evalkey.h:71]: size_t

File: src/binfhe/include/rgsw-cryptoparameters.h [hunk 36]
```diff
@@ -0,0 +1,318 @@
...
+            NativeInteger(3) * (q >> 3),  // NAND
+            NativeInteger(5) * (q >> 3),  // XOR_FAST
+            NativeInteger(1) * (q >> 3)   // XNOR_FAST
+        };
+
+        // Computes polynomials X^m - 1 that are needed in the accumulator for the
+        // CGGI bootstrapping
+        if (m_method == GINX) {
+            // loop for positive values of m
+            for (uint32_t i = 0; i < N; i++) {
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-cryptoparameters.h:138]: size_t

File: src/binfhe/include/rgsw-cryptoparameters.h [hunk 37]
```diff
@@ -0,0 +1,318 @@
...
+
+        // Precomputes a polynomial for MSB extraction
+        m_polyParams = std::make_shared<ILNativeParams>(2 * N, Q, rootOfUnity);
+        m_digitsG    = (uint32_t)std::ceil(log(Q.ConvertToDouble()) / log(static_cast<double>(m_baseG)));
+        if (m_method == AP) {
+            uint32_t digitCountR =
+                (uint32_t)std::ceil(log(static_cast<double>(q.ConvertToInt())) / log(static_cast<double>(m_baseR)));
+            // Populate digits
+            NativeInteger value = 1;
+            for (uint32_t i = 0; i < digitCountR; i++) {
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-cryptoparameters.h:93]: use size_t for loop indices

File: src/binfhe/include/rgsw-evalkey.h [hunk 38]
```diff
@@ -0,0 +1,157 @@
...
+
+    std::string SerializedObjectName() const {
+        return "RingGSWEvalKey";
+    }
+    static uint32_t SerializedVersion() {
+        return 1;
+    }
+
+private:
+    std::vector<std::vector<NativePoly>> m_elements;
```
Comment (dsuponitskiy) [inline src/binfhe/include/rgsw-evalkey.h:152]: in-class initialization

**Review by dsuponitskiy — APPROVED** [review]
Looks good to me

**Pull request merged**